### PR TITLE
correct copyright header years of code files per git data

### PR DESCRIPTION
**Pull request by phrohdoh**
<details><summary>how was this done?</summary>
years determined with the following shell command:

    git ls-tree -r --name-only HEAD | while read f; do most_recent_creation_date="$(git log --diff-filter=A --follow --format='%aI' -- $f)" ; most_recent_mod_date="$(git log -1 --format="%aI" -- $f)" ; most_recent_creation_year="$(echo $most_recent_creation_date | cut -d'-' -f1)" ; most_recent_mod_year="$(echo $most_recent_mod_date | cut -d'-' -f1)" ; echo "${most_recent_creation_year}-${most_recent_mod_year} $f" ; done

files updated via the following [deno] script:

```typescript
#!/usr/bin/env -S deno -q run --allow-read --allow-write

// ```shell
// git ls-tree -r --name-only HEAD | while read f; do most_recent_creation_date="$(git log --diff-filter=A --follow --format='%aI' -- $f)" ; most_recent_mod_date="$(git log -1 --format="%aI" -- $f)" ; most_recent_creation_year="$(echo $most_recent_creation_date | cut -d'-' -f1)" ; most_recent_mod_year="$(echo $most_recent_mod_date | cut -d'-' -f1)" ; echo "${most_recent_creation_year}-${most_recent_mod_year} $f" ; done | this-script.ts
// ```

import { readLines } from "https://deno.land/std@0.83.0/io/bufio.ts";

const allowListFileExts = [
    'cs',
    'lua',
];

for await (const line of readLines(Deno.stdin)) {
  if (!line) {
    continue;
  }

  const [yearRange, filePath] = line.split(' ');

  if (!yearRange || !filePath) {
      continue;
  }

  const ext = filePath.split('.', 2)[1];
  if (!ext || !allowListFileExts.includes(ext)) {
      continue;
  }

  const fileContent = await Deno.readTextFile(filePath);
  const newFileContent = fileContent.replace(
      /Copyright \d{4}-\d{4} The OpenRA/,
      `Copyright ${yearRange} The OpenRA`,
  );

  await Deno.writeTextFile(filePath, newFileContent);
}
```

</details>

---

Any thoughts on updating the latest year for these files to 2021 since with this PR they're all being modified?

[deno]: https://deno.land/

### CODE REVIEW

**Comment by pchote** [comment]
How about we remove the year completely? Touching every file in the repo each year is silly, and, as this PR demonstrates, the claimed years have always been wrong anyway.

**Comment by pchote** [comment]
Further musing: It would be better still if we could actually correctly list the copyrightable dates in the file headers. This would require a first automated pass that sets a list of years based on the commit history, and then a CI test that checks if the year of the commit being tested matches the last year listed in the header.

**Comment by matjaeck** [comment]
<details>
<summary>Some thoughts but IANAL and not experienced in licencing, so take at own risk.</summary>

I think these are the relevant parts from the licence:

 `For both users' and authors' sake, the GPL requires that modified versions be marked as changed, so that their problems will not be attributed erroneously to authors of previous versions.`

 `The work must carry prominent notices stating that you modified it, and giving a relevant date.`

This requires:
- Naming copyright holders (authors)
  - Copyright-Claim is suggested by the FFS and implied by the GPLS itself:
    *"However, nothing other than this License grants you permission to propagate or
    modify any covered work.  These actions infringe copyright if you do
    not accept this License."*
- Associating authors with changes in a way that prevents erroneous attribution of modifications
  - This is a general requirement and also something you are allowed to specify more strictly in additional terms per 7 b), c). If the licence makes author attribution, prohibition of misrepresentation and requirements to mark modified versions as different versions an opt-in, one can assume the basic licence sets lower demands for these points.

It also means that copyright-date is not something limited to copies or downstream, for example the same motivation to prevent false attribution can occur when one part of joint copyright holders stops working without any copies being made so far. Initial contributors in an open source project may want to mark which later versions they do not want to falsely be attributed with.

To fulfill these requirements, I would suggest doing the following:

- The referenced AUTHORS file should include everyone who has made modifications rather than give credits to important modifications. This helps contributors to enforce their copyright and strengthens the licence.
- Mark for each author the year(s) or the time-period(s) they have been doing modifications in the AUTHORS file: Tanya Adams (1996 - 2000) 
- Add the year of the last modification in the copyright header (as you suggested) to give a basic understanding who holds copyright on modifications from which time-frames

Requests like #15640 may be conflicting with your own licence terms if you have merged modifications and it may mean that a removal from AUTHORS could only happen when they transfer the copyright on their work to somebody else at the same time.

</details>

**Comment by Mailaender** [comment]
> Any thoughts on updating the latest year for these files to 2021 since with this PR they're all being modified?

I think they should update to the year they have last been modified. I doubt all files have been touched in 2020.

File: OpenRA.Mods.Cnc/FileFormats/BlowfishKeyProvider.cs [hunk 1]
```diff
@@ -1,6 +1,6 @@
 #region Copyright & License Information
 /*
- * Copyright 2007-2020 The OpenRA Developers (see AUTHORS)
+ * Copyright 2009-2020 The OpenRA Developers (see AUTHORS)
```
Comment (Mailaender) [inline OpenRA.Mods.Cnc/FileFormats/BlowfishKeyProvider.cs:3]: Last time this was modified was the last automated copyright year header change. https://github.com/OpenRA/OpenRA/commit/23b3c237b7071fd308c4664b0b6c5d719c0f3c74#diff-6ead9da50ccf0af6b203a7393c0b3c41a29ae2c82f338934380eb3d9257c8169 That should be excluded as it is a non-copyrightable change.

**Comment by pchote** [comment]
> I think they should update to the year they have last been modified. I doubt all files have been touched in 2020.

I found https://www.gnu.org/prep/maintain/html_node/Copyright-Notices.html, which recommends otherwise:

> When you add the new year, it is not required to keep track of which files have seen significant changes in the new year and which have not. It is recommended and simpler to add the new year to all files in the package, and be done with it for the rest of the year.

**Comment by Mailaender** [comment]
See https://www.eclipse.org/projects/handbook/#ip-copyright-headers which is more precise, but more work and also more prone to error and rebase conflicts. There is even https://wiki.eclipse.org/Development_Resources/How_to_Use_Eclipse_Copyright_Tool to help developers with it.

**Pull request closed without merging**